odl/trafos/convolution.py

```diff
@@ -0,0 +1,804 @@
+        return self._axes
+
+    @property
+    def kernel(self):
+        """Real-space kernel if used, else `None`.
+
+        Note that this is the scaled version ``kernel * cell_volume``
+        if ``scale=True`` was specified. Scaling here is more efficient
+        than scaling the result.
+        """
+        return self._kernel
+
+    @property
+    def kernel_space(self):
+        """Space of the convolution kernel."""
+        return getattr(self.kernel, 'space', None)
+
+    @property
+    def kernel_transform(self):
+        """Fourier-space kernel if used, else `None`.
+
+        Note that this is the scaled version
+        ``kernel_ft * (2*pi) ** (ndim/2)`` of the FT of the input
+        kernel if ``scale=True`` was specified. Scaling here is more
+        efficient than scaling the result.
+        """
+        return self._kernel_transform
+
+    def _call(self, x, out, **kwargs):
+        """Implement ``self(x, out[, **kwargs])``.
+
+        Keyword arguments are passed on to the transform.
+        """
+        if self.kernel is not None:
+            # Scipy based convolution
+            out[:] = signal.convolve(x, self.kernel, mode='same')
+        elif self.kernel_transform is not None:
+            # Convolution based on our own transforms
+            if self.domain.field == ComplexNumbers():
+                # Use out as a temporary, has the same size
+                # TODO: won't work for CUDA
+                tmp = self.transform.range.element(out.asarray())
+            else:
+                # No temporary since out has reduced size (halfcomplex)
+                tmp = None
+
+            x_trafo = self.transform(x, out=tmp, **kwargs)
+            x_trafo *= self.kernel_transform
+
+            self.transform.inverse(x_trafo, out=out, **kwargs)
+        else:
+            raise RuntimeError('both kernel and kernel_transform are None.')
+
+    def _adj_kernel(self, kernel, axes):
+        """Return adjoint kernel with adapted shape."""
+        kernel = np.asarray(kernel).conj()
+        extra_dims = self.domain.ndim - kernel.ndim
+
+        if extra_dims == 0:
+            slc = [slice(None, None, -1)] * self.domain.ndim
+            return kernel[slc]
+        else:
+            if len(axes) != extra_dims:
+                raise ValueError('kernel dim ({}) + number of axes ({}) '
+                                 'does not add up to the space dimension '
+                                 '({}).'.format(kernel.ndim, len(axes),
+                                                self.domain.ndim))
+
+            # Sparse kernel (less dimensions), blow up
+            slc = [None] * self.domain.ndim
+            for ax in axes:
+                slc[ax] = slice(None, None, -1)
+            return kernel[slc]
+
+    @property
+    def adjoint(self):
+        """Adjoint operator."""
+        if self.kernel is not None:
+            # TODO: this could be expensive. Move to init?
```

Comment:
Leave for now. We can profile later. Moving it to init also carries a space cost.